streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.kafka.streams.kstream.SessionMerger;
+import org.apache.kafka.streams.kstream.SessionWindows;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.SessionStore;
+
+import java.util.ArrayList;
+import java.util.List;
+
+class KStreamSessionWindowAggregate<K, V, T> implements KStreamAggProcessorSupplier<K, Windowed<K>, V, T> {
+
+    private final String storeName;
+    private final SessionWindows windows;
+    private final Initializer<T> initializer;
+    private final Aggregator<K, V, T> aggregator;
+    private final SessionMerger<K, T> sessionMerger;
+
+    private boolean sendOldValues = false;
+
+    KStreamSessionWindowAggregate(final SessionWindows windows,
+                                  final String storeName,
+                                  final Initializer<T> initializer,
+                                  final Aggregator<K, V, T> aggregator,
+                                  final SessionMerger<K, T> sessionMerger) {
+        this.windows = windows;
+        this.storeName = storeName;
+        this.initializer = initializer;
+        this.aggregator = aggregator;
+        this.sessionMerger = sessionMerger;
+    }
+
+    @Override
+    public Processor<K, V> get() {
+        return new KStreamSessionWindowAggregateProcessor();
+    }
+
+    @Override
+    public void enableSendingOldValues() {
+        sendOldValues = true;
+    }
+
+    private class KStreamSessionWindowAggregateProcessor extends AbstractProcessor<K, V> {
+
+        private SessionStore<K, T> store;
+        private TupleForwarder<Windowed<K>, T> tupleForwarder;
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(ProcessorContext context) {
+            super.init(context);
+            store = (SessionStore<K,  T>) context.getStateStore(storeName);
+            tupleForwarder = new TupleForwarder<>(store, context, new ForwardingCacheFlushListener<K, V>(context, sendOldValues), sendOldValues);
+        }
+
+        @Override
+        public void process(final K key, final V value) {
+            // if the key is null, we do not need proceed aggregating
+            // the record with the table
+            if (key == null) {
+                return;
+            }
+
+            final long timestamp = context().timestamp();
+            final List<KeyValue<Windowed<K>, T>> merged = new ArrayList<>();
+            final TimeWindow newTimeWindow = new TimeWindow(timestamp, timestamp);
+            TimeWindow mergedWindow = newTimeWindow;
+            T agg = initializer.apply();
+
+            try (final KeyValueIterator<Windowed<K>, T> iterator = store.findSessionsToMerge(key, timestamp - windows.gap() + 1,
```

Comment:
The javadoc is correct.
It is to do with the way i've interpreted what to do with the gap ( which is possibly confusing).
i.e, if the gap is 5 and 5 seconds has elapsed it should be a new Session. So given a timestamp of 15 i only want to find sessions where earliestEndTime >= 11 and the latestStartTime <= 19